File: touchstone/convert4to2.m
% Manual conversion from 4-port differential S-parameters to 2-port single-ended S-parameters
% No RF Toolbox needed

inputFile  = 'C2M__Z100_IL14_WC_BOR_H_L_H_THRU.s4p';   % Input differential S-parameter file
outputFile = 'channel_two_port.s2p';   % Output single-ended file

% --- Read .s4p file ---
fid = fopen(inputFile);
lines = {};
while ~feof(fid)
    line = strtrim(fgetl(fid));
    if startsWith(line, '!') || isempty(line)
        continue;
    end
    lines{end+1} = line; %#ok<SAGROW>
end
fclose(fid);

% Extract frequency and S-parameters
header = split(lines{1});
freq_unit = header{2}; % e.g. 'GHz'

data = [];
for i = 2:length(lines)
    nums = str2double(split(lines{i}));
    data = [data; nums]; %#ok<AGROW>
end

freq = data(:,1);
S = zeros(4,4,length(freq));  % 4x4 S-matrix

% Parse each line of magnitude/angle data
% Touchstone order: S11 S12 S13 S14 S21 S22 ... S44
for k = 1:length(freq)
    idx = 2;
    for r = 1:4
        for c = 1:4
            mag = data(k,idx); ang = data(k,idx+1);
            S(r,c,k) = mag .* exp(1j*ang*pi/180);
            idx = idx + 2;
        end
    end
end

% --- Perform mixed-mode to single-ended conversion ---
M = (1/sqrt(2)) * [1  0  1  0;
                   -1 0  1  0;
                    0  1  0  1;
                    0 -1  0  1];

S2 = zeros(2,2,length(freq));

for k = 1:length(freq)
    S2(:,:,k) = M * S(:,:,k) * inv(M);
    % Take the 2x2 differential block
    S2(:,:,k) = S2(1:2,1:2);
end

% --- Write output .s2p file ---
fid = fopen(outputFile, 'w');
fprintf(fid, '# %s S MA R 50\n', freq_unit);
for k = 1:length(freq)
    fprintf(fid, '%e ', freq(k));
    for r = 1:2
        for c = 1:2
            mag = abs(S2(r,c,k));
            ang = angle(S2(r,c,k))*180/pi;
            fprintf(fid, '%e %e ', mag, ang);
        end
    end
    fprintf(fid, '\n');
end
fclose(fid);

fprintf('Converted %s → %s\n', inputFile, outputFile);

File: touchstone/convert4to2.py
import skrf as rf
import numpy as np
import sys

def convert_differential_to_single_ended(input_file, output_file):
    """
    Convert a differential 4-port S-parameter file (S4P)
    into a 2-port differential (S2P) Touchstone file.

    Expected port mapping:
        Port 1 -> TX+
        Port 2 -> TX-
        Port 3 -> RX+
        Port 4 -> RX-
    """

    # Load the 4-port S-parameter file
    s4p = rf.Network(input_file)
    if s4p.number_of_ports != 4:
        raise ValueError("Input file must have 4 ports (differential pair system).")

    # Convert single-ended (SE) -> mixed-mode (MM)
    # New API in scikit-rf >= 0.25
    mm = rf.network.se2gmm(s4p)

    # Extract the differential-to-differential (Sdd) 2x2 block
    # Mixed-mode order: [diff1, diff2, comm1, comm2]
    Sdd11 = mm.s[:, 0, 0]
    Sdd12 = mm.s[:, 0, 1]
    Sdd21 = mm.s[:, 1, 0]
    Sdd22 = mm.s[:, 1, 1]

    # Construct a 2-port differential S-matrix
    s2p_matrix = np.zeros((len(Sdd11), 2, 2), dtype=complex)
    s2p_matrix[:, 0, 0] = Sdd11
    s2p_matrix[:, 0, 1] = Sdd12
    s2p_matrix[:, 1, 0] = Sdd21
    s2p_matrix[:, 1, 1] = Sdd22

    # Create the new 2-port network
    s2p = rf.Network(frequency=s4p.frequency, s=s2p_matrix, z0=s4p.z0[0, 0])

    # Write the resulting S2P file
    s2p.write_touchstone(output_file)
    print(f"✅ Successfully converted {input_file} → {output_file}.s2p")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python convert4to2.py input.s4p output_basename")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2]
    convert_differential_to_single_ended(input_file, output_file)
